File: FrontEnd.Mobile/lib/core/correlation/correlation_id.dart
import 'package:uuid/uuid.dart';

/// Aligns with backend [X-Correlation-Id] middleware and web client.
abstract final class CorrelationHeaders {
  static const headerName = 'X-Correlation-Id';
}

const _uuid = Uuid();

String createCorrelationId() => _uuid.v4().replaceAll('-', '');

String? _lastCorrelationId;

String? get lastCorrelationId => _lastCorrelationId;

void setLastCorrelationId(String? id) {
  _lastCorrelationId = id;
}

String? readCorrelationIdFromHeaders(Map<String, List<String>>? headers) {
  if (headers == null) return null;
  for (final key in headers.keys) {
    if (key.toLowerCase() == CorrelationHeaders.headerName.toLowerCase()) {
      final values = headers[key];
      if (values != null && values.isNotEmpty && values.first.isNotEmpty) {
        return values.first;
      }
    }
  }
  return null;
}

File: FrontEnd/apps/web/src/shared/errors/correlationId.ts
/**
 * Correlation ID helpers — aligned with backend X-Correlation-Id middleware.
 */

export const CORRELATION_HEADER = 'X-Correlation-Id';

let lastCorrelationId: string | null = null;

export function createCorrelationId(): string {
  return crypto.randomUUID().replace(/-/g, '');
}

export function getLastCorrelationId(): string | null {
  return lastCorrelationId;
}

export function setLastCorrelationId(id: string | null): void {
  lastCorrelationId = id;
}

export function readCorrelationIdFromHeaders(
  headers: Record<string, unknown> | undefined,
): string | null {
  if (!headers) return null;
  const value =
    headers[CORRELATION_HEADER] ??
    headers[CORRELATION_HEADER.toLowerCase()] ??
    headers['x-correlation-id'];
  if (typeof value === 'string' && value.length > 0) return value;
  if (Array.isArray(value) && typeof value[0] === 'string') return value[0];
  return null;
}
